File: Pokedex/Pokedex/PokemonTableViewController.m
//
//  PokemonTableViewController.m
//  Pokedex
//
//  Created by Claudia Maciel on 8/14/20.
//  Copyright © 2020 thecoderpilot. All rights reserved.
//

#import "PokemonTableViewController.h"
#import "Pokemon.h"
#import "Pokedex-Swift.h"
#import "PokemonDetailViewController.h"

@interface PokemonTableViewController ()

@property (nonatomic) NSArray<Pokemon *> *pokemon;

@end

@implementation PokemonTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self updateViews];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.pokemon.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"PokemonCell" forIndexPath:indexPath];
    
    cell.textLabel.text = self.pokemon[indexPath.row].name;
    
    return cell;
}


// MARK: - Functions
- (void)updateViews
{
    [PokemonController.sharedController fetchAllPokemonWithCompletion:^(NSArray<Pokemon *> * _Nullable pokemon, NSError * _Nullable error) {
        if (error) {
            NSLog(@"Failed to fetch pokemon with error: %@", error);
            return;
        }

        if (pokemon) {
            self.pokemon = pokemon;
            dispatch_async(dispatch_get_main_queue(), ^(void){
                [self.tableView reloadData];
            });
        } else {
            NSLog(@"Pokemon array was nil");
            return;
        }
    }];
}



#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if([segue.identifier isEqualToString:@"PokemonDetailSegue"])
    {
        PokemonDetailViewController *pokemonDetailVC = segue.destinationViewController;
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        pokemonDetailVC.pokemon = [self.pokemon objectAtIndex:indexPath.row];
    }
}


@end

File: Pokedex/Pokedex/PokemonTableViewController.swift
//
//  PokemonTableViewController.swift
//  Pokedex
//
//  Created by Enayatullah Naseri on 6/12/20.
//  Copyright © 2020 Enayatullah Naseri. All rights reserved.
//

import UIKit

class PokemonTableViewController: UITableViewController {
    
    // MARK: - Properties
    var pokemons = [Pokemon]()
    
    // MARK: - ViewDidLoad
    override func viewDidLoad() {
        super.viewDidLoad()
        fetchPokemon()
    }

    // MARK: - Count
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete implementation, return the number of rows
        return pokemons.count
    }

    // MARK: - Cell
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "PokemonCell", for: indexPath)

        let pokemon = pokemons[indexPath.row]
        cell.textLabel?.text = pokemon.pokemonName.capitalized

        return cell
    }
    
        // MARK: - Fetching data - function
    //    let errorFetch =
        private func fetchPokemon() {
            
            PokemonController.shared.fetchAllPokemon { (pokemon, error) in
                
                if let error = error {
                    print("Error fetching: \(error)")
                    return
                }
                
                guard let pokemon = pokemon else {
                    print("Couldnt find a Pokemon")
                    return
                }
                
                self.pokemons.append(contentsOf: pokemon)
                
                DispatchQueue.main.async {
                    self.tableView.reloadData()
                }
            }
        }

    
    // MARK: - Navigation
    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if segue.identifier == "ShowPokemonDetailSegue" {
            guard let destinationVC = segue.destination as? PokemonDetailViewController,
                let indexPath = tableView.indexPathForSelectedRow else { return }
            destinationVC.pokemon = pokemons[indexPath.row]
        }
        
    }

}
